src/persist-client/src/internal/state_versions.rs

```diff
@@ -620,19 +625,87 @@ impl StateVersions {
             "add_and_remove_rollups should apply to the empty state"
         );
 
-        let rollup = self.encode_rollup_blob(shard_metrics, &initial_state, rollup.key);
+        let rollup = self.encode_rollup_blob(shard_metrics, &initial_state, vec![], rollup.key);
         let () = self.write_rollup_blob(&rollup).await;
         assert_eq!(initial_state.seqno, rollup.seqno);
 
         let diff = StateDiff::from_diff(&empty_state.state, &initial_state.state);
         (initial_state, diff)
     }
 
-    /// Encodes the given state as a rollup to be written to the specified key.
+    pub async fn write_rollup_for_state<K, V, T, D>(
+        &self,
+        shard_metrics: &ShardMetrics,
+        state: &TypedState<K, V, T, D>,
+        rollup_id: &RollupId,
+    ) -> Option<EncodedRollup>
+    where
+        K: Debug + Codec,
+        V: Debug + Codec,
+        T: Timestamp + Lattice + Codec64,
+        D: Semigroup + Codec64,
+    {
+        let (latest_rollup_seqno, _rollup) = state.latest_rollup();
+        let seqno = state.seqno();
+
+        let diffs: Vec<_> = self
+            .fetch_all_live_diffs_gt_seqno::<K, V, T, D>(&state.shard_id, *latest_rollup_seqno)
+            .await;
+
+        match diffs.first() {
+            None => {
+                // early-out because these are no diffs past our latest rollup.
+                //
+                // this should only occur in the initial state, but we can write a more
+                // general assertion: if no live diffs exist past this state's latest
+                // known rollup, then that rollup must be for the latest known state.
+                self.metrics.state.rollup_write_noop_latest.inc();
+                assert_eq!(seqno, *latest_rollup_seqno);
+                return None;
+            }
+            Some(first) => {
+                // early-out if it is no longer possible to inline all the diffs from
```

Comment:
this shouldn't be possible, right?

Reply:
it should be - one process gets told to write a rollup, the maintenance task falls arbitrarily behind, meanwhile other processes write new rollups and GC truncates past the first process's task

Reply:
ah, yeah, i see! lets include that context in the comment, please!